Added a new Spectrogram layer based on Conv1D operations, supporting GPU-parallelization and fine-tuning

mostafa-mahmoud opened the pull request:
Added a Spectrogram layer that computes several modes of the Spectrograms (like the magnitude, log-magnitude, power spectral density). The computation aims to serve as an official standardized layer for computing spectrograms while being part of the model. The computations are based on Conv1D operations, which makes it parallelizable when running the model on GPUs. Also, since the computations are based on trainable kernels (which can be adjusted to be trainable or not like any other kernel), further fine-tuning of the Spectrogram computations is possible.

fchollet commented on keras/src/layers/preprocessing/spectrogram.py:
```diff
@@ -0,0 +1,325 @@
+        elif self.mode == "log":
+            return ops.log(backend.epsilon() + linear_stft)
+
+        raise NotImplementedError(f"{self.mode} mode is not implemented")
```
Input validation should happen in the constructor not at call time

mostafa-mahmoud replied:
I removed the validation in `__call__`, which was redundant since an equivalent check is already made in the constructor.

fchollet commented on keras/src/layers/preprocessing/spectrogram_test.py:
```diff
@@ -0,0 +1,196 @@
+    @pytest.mark.requires_trainable_backend
+    def test_spectrogram_error(self):
+        rnd = np.random.RandomState(41)
+        X = rnd.uniform(low=-1, high=1, size=(4, 160000, 1)).astype(np.float64)
```
Use lowercase variable names

mostafa-mahmoud replied:
Renamed all the uppercase variables in the test module to lowercase.

fchollet reviewed (commented):
Interesting feature, thanks for the PR!

What would be the main use cases for this? And what is the usage pattern in terms of what initializer to use, when to set trainable = False, etc? Can you a simple tutorial that demonstrates the value?

mostafa-mahmoud commented on keras/src/layers/preprocessing/spectrogram.py:
```diff
@@ -0,0 +1,325 @@
+
+
+@keras_export("keras.layers.Spectrogram")
+class Spectrogram(layers.Layer):
```
I renamed to `STFTSpectrogram`, which is more specific.

However, I aimed for this to be extended in later PRs to also include Mel-Spectrogram, LogMel-Spectrogram, and MFCCs. These are all audio-based spectrograms, unlike the layer I just committed which is more generic for time-series signals generally. Supporting these output modes would require extra computations at the end of the `__call__` function.

If all of these variations would be in one layer in the future, then maybe having the name `Spectrogram` is better, which will make this more generic. However, if this is too monolithic and should be handled in a new layer(maybe inheriting from the current layer), then I think the current naming `STFTSpectrogram` is sufficient.

What do you think? Should I use `STFTSpectrogram` or `Spectrogram`? (keeping in mind the possible future extension to Mel-Spectrograms and MFCCs)

mostafa-mahmoud commented:
> Interesting feature, thanks for the PR!
> 
> 
> 
> What would be the main use cases for this? And what is the usage pattern in terms of what initializer to use, when to set trainable = False, etc? Can you a simple tutorial that demonstrates the value?

This layer comes with its default initializer (the `STFTInitializer` in this pushed code), which computes STFT out-of-the-box.

The main two use-cases are:
1) Set `trainable=False`, and this is purely a preprocessing layer.
2) Set `trainable=True`, then this becomes a fine-tuned Spectrogram, which can be beneficial in some cases. Cheuk et. al. 2020 study this.

I can craft an example using this layer.

There are mainly three reasons for this this contribution:

1. Standardize the spectrogram preprocessing as part of the model. Also, this allows deploying the model directly without taking the preprocessing code to the server.

I often had issues with different implementations across different libraries.

The convenience of this layer can make it a standard practice.

2. Using implementation based on convolutions, which makes the operation faster on GPUs due to parallelization, compared to the standard FFT. It also resolves some bottlenecks of preprocessing the Spectrogram on CPU before transferring it to GPU.

All of these allow faster execution.

3. The use case of `trainable=True`, which could lead to some improvements.

---

K. W. Cheuk, H. Anderson, K. Agres and D. Herremans, "nnAudio: An on-the-Fly GPU Audio to Spectrogram Conversion Toolbox Using 1D Convolutional Neural Networks," in IEEE Access, vol. 8, pp. 161981-162003, 2020.

mostafa-mahmoud commented:
@fchollet I added another use case of outputting 2D images as well as the original 1D time signals. I also added a code example as you mentioned, you will find the PR on `keras-io` [here](https://github.com/keras-team/keras-io/pull/1947).

fchollet commented on keras/src/layers/preprocessing/stft_spectrogram.py:
```diff
@@ -0,0 +1,410 @@
+
+
+class STFTInitializer(initializers.Initializer):
+    def __init__(self, side, window, scaling, periodic=None):
```
Is `periodic` supposed to be boolean? Why is it `None` rather than `False`?

mostafa-mahmoud replied:
This was a mistake, made it to `False` by default.

fchollet commented on keras/src/layers/preprocessing/stft_spectrogram.py:
```diff
@@ -0,0 +1,410 @@
+from keras.src.utils.module_utils import scipy
+
+
+class STFTInitializer(initializers.Initializer):
```
Please add a docstring with an `Args:` section. Also please move the class to `initializers/constant_initializers.py` (since it isn't random)

mostafa-mahmoud replied:
Moved it, updated the docstrings, and added a unittest for it.

fchollet commented on keras/src/layers/preprocessing/stft_spectrogram.py:
```diff
@@ -0,0 +1,410 @@
+
+@keras_export("keras.layers.STFTSpectrogram")
+class STFTSpectrogram(layers.Layer):
+    """
```
Add a one-line description here.

fchollet commented on keras/src/layers/preprocessing/stft_spectrogram.py:
````diff
@@ -0,0 +1,410 @@
+        the number of FFT bins `(fft_length // 2 + 1)`.
+
+    Example:
+        ```python
````
Unindent (same level as docstring)

fchollet commented on keras/src/layers/preprocessing/stft_spectrogram.py:
````diff
@@ -0,0 +1,410 @@
+        )
+        output = spectrogram_layer(input_signal)
+        ```
+
````
Since this layer is fairly niche, please illustrate its usage with detailed code examples (with descriptions of what they do and what they'd be used for), e.g. "different modes of output", "non trainable init vs fine-tuning", etc.

The user should be able to read the docstring and understand: what is this layer for? when would I need it? how do I use it?

mostafa-mahmoud replied:
I added now three code examples demonstrating the different use cases. Also check the more comprehensive tutorial I added [here](https://github.com/keras-team/keras-io/pull/1947).

fchollet reviewed (commented):
Thanks for the update!

fchollet reviewed (approved):
Looking good -- thank you for the contribution!
